# Примеры и упражнения по Python3  - средний уровень - часть 1

Автор-составитель - Михаил Колодин

Версия 2021-04-16 от 2021-07-21 1.5

Разделы:
* [Циклы](#loops)
* [Строки](#strings)
* [Кортежи и списки, словари и множества](#structs1)
    * [Красивые матрицы](#structs2)
* [Функции](#functions)
* [Исключения](#except)
* [Математика](#math)

---
**Циклы** <a name=loops></a>

---
***Факториал $n! = 1 \cdot 2 \cdot 3 \cdot \dots \cdot n$***

In [1]:
UP = 30
f = 1

In [2]:
for i in range(1, UP+1):
    f *= i             #  то же, что f = f * i
    print("%10i!  =" % i, "%40i" % f)

         1!  =                                        1
         2!  =                                        2
         3!  =                                        6
         4!  =                                       24
         5!  =                                      120
         6!  =                                      720
         7!  =                                     5040
         8!  =                                    40320
         9!  =                                   362880
        10!  =                                  3628800
        11!  =                                 39916800
        12!  =                                479001600
        13!  =                               6227020800
        14!  =                              87178291200
        15!  =                            1307674368000
        16!  =                           20922789888000
        17!  =                          355687428096000
        18!  =                         640237370

---
***Сколько нулей?***

__Вопрос.__ Как видно, n! быстро растёт, и на конце этого большого числа появляется много нулей. 
А сколько именно?
Само число n! вычислять не надо.

__Решение.__ Разложим n! на простые множители. 
Нуль в конце появляется тогда и только тогда, когда перемножаются 2 и 5. 
Посчитаем пары двоек и пятёрок. 
Пятёрок меньше, так что посчитаем их. 
Сумма и даст число нулей на конце n!.

In [14]:
N = 100    # это исходное число, N!

nuls = 0
for x in range(2, N+1):
    n = x
    while n % 5 == 0:
        n //= 5
        nuls += 1
        
print(f"на конце числа {N}! будет нулей: {nuls}.")

на конце числа 100! будет нулей: 24.


---
***Константа е - аппроксимация***

См. [статью в википедии](https://ru.wikipedia.org/wiki/E_(%D1%87%D0%B8%D1%81%D0%BB%D0%BE))

$$e = \sum\limits_{n=0}^\infty \frac1{n!} = \frac1{1!} + \frac1{2!} + \frac1{3!} + \frac1{4!} + \frac1{5!} + ... = \frac11 + \frac12 + \frac16 + \frac1{24} + \frac1{120} + ... \approx 2,7182818284590452353602874713527…$$

In [34]:
UP = 30
add = 1
sum = 1

In [35]:
for i in range(1, UP):
    add /= i
    sum += add
    print("%30.25g -> %30.25g" % (add, sum))

                             1 ->                              2
                           0.5 ->                            2.5
   0.1666666666666666574148081 ->      2.66666666666666651863693
  0.04166666666666666435370203 ->      2.70833333333333303727386
 0.008333333333333333217685102 ->     2.716666666666666341001246
 0.001388888888888888941894328 ->     2.718055555555555447000415
0.0001984126984126984125263171 ->     2.718253968253968366752815
2.480158730158730156578964e-05 ->     2.718278769841270037233016
2.755731922398589251095059e-06 ->     2.718281525573192247691168
2.755731922398589356974178e-07 ->     2.718281801146384513145904
2.505210838544172022386618e-08 ->     2.718281826198492900914516
2.087675698786810018655515e-09 ->     2.718281828286168710917536
1.605904383682161592580233e-10 ->     2.718281828446759362805096
 1.14707455977297261228837e-11 ->     2.718281828458230187095523
7.647163731819817415255803e-13 ->     2.718281828458994908714885
4.77947733238738588453487

---
**Строки** <a name=strings></a>

---
***Шифр Цезаря***

In [1]:
# setup shift
shift = 1
# setup method
import string
alfa = string.digits + string.ascii_letters
alfa_len = len(alfa)
print(alfa_len, alfa)

62 0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ


In [18]:
sopen = "Hello commander! Start at noon! Ceasar"
#sopen = "Hello commander! Start at noon! Ceasar zZ"

#sopen = "Здравствуй, Цезарь! Идущие на смерть тебя приветствуют!"
#sopen = "Ave, Caesar, morituri te salutant"
#sopen = "Николай неохотно поплёлся к директору: «Ave, Caesar, morituri te salutant!» — воскликнул он в дверях."

shift = 3
print(sopen)

Hello commander! Start at noon! Ceasar


In [19]:
scoded = "".join(alfa[(alfa.index(c) + shift) % alfa_len] if c in alfa else c for c in sopen)
print(scoded)

Khoor frppdqghu! Vwduw dw qrrq! Fhdvdu


In [20]:
sback = "".join(alfa[(alfa.index(c) - shift) % alfa_len] if c in alfa else c for c in scoded)
print(sback)

Hello commander! Start at noon! Ceasar


___расширение на несколько языков___

In [6]:
# setup shift
shift = 1
# setup method
import string
alfa_dig = string.digits
alfa_enlo = string.ascii_lowercase
alfa_enup = string.ascii_uppercase
alfa_rulo = "абвгдеёжзийклмнопрстуфхцчшщъыьэюя"
alfa_ruup = "АБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ"
alfas = alfa_dig, alfa_enlo, alfa_enup, alfa_rulo, alfa_ruup
print(alfas)

('0123456789', 'abcdefghijklmnopqrstuvwxyz', 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'абвгдеёжзийклмнопрстуфхцчшщъыьэюя', 'АБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ')


In [21]:
def tuda(sin, shift):
    """ шифрование текста """
    if not sin or shift==0:
        return sin
    return "".join([bukva(c, shift) for c in sin])
    
def bukva(cin, shift):
    """ обрабатываем 1 букву """
    for alfa in alfas:
        if cin in alfa:
            return alfa[(alfa.index(cin) + shift) % len(alfa)]
    return cin

def obratno(sin, shift):
    """ обратное шифрование"""
    return tuda(sin, -shift)

In [22]:
#sopen = "Hello commander! Start at noon! Ceasar"
#sopen = "Hello commander! Start at noon! Ceasar zZ"
#sopen = "Здравствуй, Цезарь! Идущие на смерть тебя приветствуют!"
#sopen = "Ave, Caesar, morituri te salutant"
sopen = "123. Николай неохотно поплёлся к директору: «Ave, Caesar, morituri te salutant!» — воскликнул он в дверях."

shift = 3
print(sopen)
scoded = tuda(sopen, shift)
print(scoded)
sback = obratno(scoded, shift)
print(sback)

123. Николай неохотно поплёлся к директору: «Ave, Caesar, morituri te salutant!» — воскликнул он в дверях.
456. Рлнсогм рзсшсхрс тстоиофв н жлузнхсуц: «Dyh, Fdhvdu, prulwxul wh vdoxwdqw!» — есфнолнрцо ср е жезувш.
123. Николай неохотно поплёлся к директору: «Ave, Caesar, morituri te salutant!» — воскликнул он в дверях.


---
***Подсчёт символов***

(найти 10 самых частых букв)

In [22]:
stroka1 = "Twas brillig, and the slithy toves Did gyre and gimble in the wabe: All mimsy were the borogoves, And the mome raths outgrabe."

stroka2 = "Варкалось. Хливкие шорьки Пырялись по наве, И хрюкотали зелюки, Как мюмзики в мове."

# "Гло́кая ку́здра ште́ко будлану́ла бо́кра и курдя́чит бокрёнка"
stroka3 = "Глокая куздра штеко будланула бокра и курдячит бокрёнка"

# панграмма
stroka4 = "Съешь ещё этих мягких французских булок да выпей же чаю"

stroka5 = "A Counter is a container that keeps track of how many times equivalent values are added. It can be used to implement the same algorithms for which bag or multiset data structures are commonly used in other languages."\
    "Counter supports three forms of initialization. Its constructor can be called with a sequence of items, a dictionary containing keys and counts, or using keyword arguments mapping string names to counts."\
    "This example counts the letters appearing in all of the words in the system dictionary to produce a frequency distribution, then prints the three most common letters. Leaving out the argument to most_common() produces a list of all the items, in order of frequency."

stroka6 = "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."

# Гомер. Одиссея
stroka7 = """
Муза, скажи мне о том многоопытном муже, который,
Странствуя долго со дня, как святой Илион им разрушен,
Многих людей города посетил и обычаи видел,
Много и сердцем скорбел на морях, о спасенье заботясь
Жизни своей и возврате в отчизну сопутников; тщетны
Были, однако, заботы, не спас он сопутников: сами
Гибель они на себя навлекли святотатством, безумцы,
Съевши быков Гелиоса, над нами ходящего бога, —
День возврата у них он похитил. Скажи же об этом
Что-нибудь нам, о Зевесова дочь, благосклонная Муза.
Все уж другие, погибели верной избегшие, были
Дома, избегнув и брани и моря; его лишь, разлукой
С милой женой и отчизной крушимого, в гроте глубоком
Светлая нимфа Калипсо, богиня богинь, произвольной
Силой держала, напрасно желая, чтоб был ей супругом.
Но когда, наконец, обращеньем времен приведен был
Год, в который ему возвратиться назначили боги
В дом свой, в Итаку (но где и в объятиях верных друзей он
Всё не избег от тревог), преисполнились жалостью боги
Все; Посейдон лишь единый упорствовал гнать Одиссея,
Богоподобного мужа, пока не достиг он отчизны."""

stroka = stroka4

print(stroka)

Съешь ещё этих мягких французских булок да выпей же чаю


In [23]:
# не по-питоновски (ну, частично)
HOWMANY = 10    # сколько букв искать?
cnt = {}

# counting all characters
for c in stroka:
    cc = c.lower()
    if cc not in cnt:
        cnt[cc] = 0
    cnt[cc] += 1
        
# finding max by sorting
arr = [(k, v) for k, v in cnt.items()]
print("unsorted:", arr)

arr.sort(key = lambda x: x[1], reverse = True)
print("sorted:", arr)

loc = [x[0] for x in arr[:HOWMANY]]

print("\n* most often characters are:", loc)

unsorted: [('с', 2), ('ъ', 1), ('е', 4), ('ш', 1), ('ь', 1), (' ', 9), ('щ', 1), ('ё', 1), ('э', 1), ('т', 1), ('и', 3), ('х', 3), ('м', 1), ('я', 1), ('г', 1), ('к', 3), ('ф', 1), ('р', 1), ('а', 3), ('н', 1), ('ц', 1), ('у', 2), ('з', 1), ('б', 1), ('л', 1), ('о', 1), ('д', 1), ('в', 1), ('ы', 1), ('п', 1), ('й', 1), ('ж', 1), ('ч', 1), ('ю', 1)]
sorted: [(' ', 9), ('е', 4), ('и', 3), ('х', 3), ('к', 3), ('а', 3), ('с', 2), ('у', 2), ('ъ', 1), ('ш', 1), ('ь', 1), ('щ', 1), ('ё', 1), ('э', 1), ('т', 1), ('м', 1), ('я', 1), ('г', 1), ('ф', 1), ('р', 1), ('н', 1), ('ц', 1), ('з', 1), ('б', 1), ('л', 1), ('о', 1), ('д', 1), ('в', 1), ('ы', 1), ('п', 1), ('й', 1), ('ж', 1), ('ч', 1), ('ю', 1)]

* most often characters are: [' ', 'е', 'и', 'х', 'к', 'а', 'с', 'у', 'ъ', 'ш']


In [24]:
# по-питоновски
import collections as coll

cnt = coll.Counter(stroka.lower())
print("* most often characters are:", [x[0] for x in cnt.most_common(HOWMANY)])

* most often characters are: [' ', 'е', 'и', 'х', 'к', 'а', 'с', 'у', 'ъ', 'ш']


---
***Анаграммы***

Анаграммы - слова (или предложения), состоящите из одних и  тех же букв, но в разном порядке.

Дан набор строк. Определить, какие из них явлеются анаграммами каких.

In [25]:
# список слов
words = "астероид кукла трос кулак каприз алкаш сорт приказ шнурок мышка осока кабан камыш поза директор коршун диаретса лунь банка ром срам шкала нуль мор марс рост остр ".split()
print(sorted(words))

['алкаш', 'астероид', 'банка', 'диаретса', 'директор', 'кабан', 'камыш', 'каприз', 'коршун', 'кукла', 'кулак', 'лунь', 'марс', 'мор', 'мышка', 'нуль', 'осока', 'остр', 'поза', 'приказ', 'ром', 'рост', 'сорт', 'срам', 'трос', 'шкала', 'шнурок']


In [26]:
# ищем анаграммы
from pprint import pprint as pp

dow = {}
for w in words:
    ws = "".join(sorted(w))
    if ws not in dow:
        dow[ws] = [w]
    else:
        dow[ws] += [w]
#pp(dow)
res = [v for v in dow.values() if len(v) > 1]
pp(res)

[['кукла', 'кулак'],
 ['трос', 'сорт', 'рост', 'остр'],
 ['каприз', 'приказ'],
 ['алкаш', 'шкала'],
 ['шнурок', 'коршун'],
 ['мышка', 'камыш'],
 ['кабан', 'банка'],
 ['лунь', 'нуль'],
 ['ром', 'мор'],
 ['срам', 'марс']]


---
**Кортежи и списки, словари и множества** <a name=structs1></a>

---
***Работа с изменяемыми значениями (mutables)***

In [20]:
m1 = [1, 2, 3]
m2 = m1
print(m1)
print(m2)

[1, 2, 3]
[1, 2, 3]


In [21]:
m1[1] = 55
print(m1)
print(m2)

[1, 55, 3]
[1, 55, 3]


Python отличается от многих других ЯП. 

В нём нет присваивания.

При "присваивании" не происходит копирования информации.

Знаком "=" обозначается операция установления (копирования) ссылки на объект в памяти.

Причём переменные живут в своих пространствах, а значения - в своей глобальной памяти (куче).

---
***Поиск самой длинной положительной последовательности***

Найти в последовательности целых чисел подпоследовательность максимальной длины, состоящую только из положительных чисел.

In [1]:
# программа
from random import randint as ri

def posseq1(seq):
    """поиск последовательности"""
    longseq = 0
    longstart = -1
    for i in range(LEN):
        if seq[i] <= 0:
            continue
        thisseq = 1
        thisstart = i
        for j in range(i+1, LEN):
            if seq[j] <= 0:
                break
            else:
                thisseq = j - i + 1
        if thisseq > longseq:
            longseq = thisseq
            longstart = thisstart
            #print(longseq, longstart, seq[longstart : longstart + longseq])
    return longseq, longstart

In [15]:
# данные
LEN = 20
MIN = -10
MAX = +10

# проверка
seq = [ri(MIN, MAX) for _ in range(LEN)]
print(seq)

[8, 9, 7, -1, 7, 1, -5, 6, -4, 0, 1, -4, -1, 4, 0, -4, -2, 2, -8, -3]


In [16]:
print(seq)
plen, pstart = posseq1(seq)
print(f"sequence of length {plen}, starting from {pstart}, is {seq[pstart : pstart+plen]}")

[8, 9, 7, -1, 7, 1, -5, 6, -4, 0, 1, -4, -1, 4, 0, -4, -2, 2, -8, -3]
sequence of length 3, starting from 0, is [8, 9, 7]


Замечание: решение не оптимально, можно его заметно ускорить.
Найти место и ускорить... :)

Подсказка: алгоритм многократно проходит по длинным последовательностям. Это не влияет на результат, но замедляет работу.

In [5]:
# программа - попытка исправления неэффективности
from random import randint as ri

def posseq2(seq):
    """поиск последовательности"""
    longseq = 0
    longstart = -1
    skipto = -1
    for i in range(LEN):
        if seq[i] <= 0:
            continue
        if i < skipto:
            continue
        skipto = -1
        thisseq = 1
        thisstart = i
        for j in range(i+1, LEN):
            if seq[j] <= 0:
                break
            else:
                thisseq = j - i + 1
                skipto = j
        if thisseq > longseq:
            longseq = thisseq
            longstart = thisstart
            #print(longseq, longstart, seq[longstart : longstart + longseq])
    return longseq, longstart

In [9]:
# данные
LEN = 20
MIN = -10
MAX = +10

# проверка
seq = [ri(MIN, MAX) for _ in range(LEN)]
print(seq)

[3, 6, 5, -7, 4, 8, 9, -1, 4, -10, 9, 6, -10, -5, 7, 8, -7, -1, 8, 8]


In [17]:
print(seq)
plen, pstart = posseq2(seq)
print(f"sequence of length {plen}, starting from {pstart}, is {seq[pstart : pstart+plen]}")

[8, 9, 7, -1, 7, 1, -5, 6, -4, 0, 1, -4, -1, 4, 0, -4, -2, 2, -8, -3]
sequence of length 3, starting from 0, is [8, 9, 7]


----
__Красивые матрицы__ <a name=structs2></a>

In [3]:
# подключаем нужное

from random import randint as ri
from pprint import pp

In [34]:
# красиво показываем
def show(a, width=5):
    """показать матрицу"""
    lena = len(a)
    print(f"matrix of size {lena}")
    for i in range(lena):
        for j in range(lena):
            print(f"{a[i][j] : {width}}", end="")
        print()
    print()

In [82]:
# красота №1 "просто нули"

def mat_1(size):
    """просто нули"""
    return [[0 for x in range(size)] for y in range(size)]

In [83]:
# показ красоты №1
show(mat_1(3))
show(mat_1(5))
show(mat_1(7))

matrix of size 3
    0    0    0
    0    0    0
    0    0    0

matrix of size 5
    0    0    0    0    0
    0    0    0    0    0
    0    0    0    0    0
    0    0    0    0    0
    0    0    0    0    0

matrix of size 7
    0    0    0    0    0    0    0
    0    0    0    0    0    0    0
    0    0    0    0    0    0    0
    0    0    0    0    0    0    0
    0    0    0    0    0    0    0
    0    0    0    0    0    0    0
    0    0    0    0    0    0    0



In [84]:
# красота №2 "две диагонали"

def mat_2(size):
    """просто нули"""
    a = [[0 for x in range(size)] for y in range(size)]
    for i in range(size):
        a[i][i] = 1
        a[i][size-i-1] = 1
        a[size-i-1][i] = 1
        a[size-i-1][size-i-1] = 1
    return a

In [85]:
# показ красоты №2
show(mat_2(3))
show(mat_2(5))
show(mat_2(7))

matrix of size 3
    1    0    1
    0    1    0
    1    0    1

matrix of size 5
    1    0    0    0    1
    0    1    0    1    0
    0    0    1    0    0
    0    1    0    1    0
    1    0    0    0    1

matrix of size 7
    1    0    0    0    0    0    1
    0    1    0    0    0    1    0
    0    0    1    0    1    0    0
    0    0    0    1    0    0    0
    0    0    1    0    1    0    0
    0    1    0    0    0    1    0
    1    0    0    0    0    0    1



In [86]:
# красота №3 "от диагонали"

def mat_3(size):
    """просто нули"""
    a = [[0 for x in range(size)] for y in range(size)]
    for i in range(size):
        for j in range(size):
            a[i][j] = abs(i-j)
    return a

In [87]:
# показ красоты №3
show(mat_3(3))
show(mat_3(5))
show(mat_3(7))

matrix of size 3
    0    1    2
    1    0    1
    2    1    0

matrix of size 5
    0    1    2    3    4
    1    0    1    2    3
    2    1    0    1    2
    3    2    1    0    1
    4    3    2    1    0

matrix of size 7
    0    1    2    3    4    5    6
    1    0    1    2    3    4    5
    2    1    0    1    2    3    4
    3    2    1    0    1    2    3
    4    3    2    1    0    1    2
    5    4    3    2    1    0    1
    6    5    4    3    2    1    0



In [88]:
# красота №4 "ямка"

def mat_4(size):
    """просто нули"""
    a = [[0 for x in range(size)] for y in range(size)]
    cen = size//2
    for r in range((size-1)//2):
        for i in range(size):
            for j in range(size):
                a[i][j] = max( abs(cen-i), abs(cen-j) )
    return a

In [89]:
# показ красоты №4
show(mat_4(3))
show(mat_4(5))
show(mat_4(7))

matrix of size 3
    1    1    1
    1    0    1
    1    1    1

matrix of size 5
    2    2    2    2    2
    2    1    1    1    2
    2    1    0    1    2
    2    1    1    1    2
    2    2    2    2    2

matrix of size 7
    3    3    3    3    3    3    3
    3    2    2    2    2    2    3
    3    2    1    1    1    2    3
    3    2    1    0    1    2    3
    3    2    1    1    1    2    3
    3    2    2    2    2    2    3
    3    3    3    3    3    3    3



In [90]:
# красота №5 "горка"

def mat_5(size):
    """просто нули"""
    a = [[0 for x in range(size)] for y in range(size)]
    cen = size//2
    for r in range((size-1)//2):
        for i in range(size):
            for j in range(size):
                a[i][j] = cen - max( abs(cen-i), abs(cen-j) )
    return a

In [91]:
# показ красоты №5
show(mat_5(3))
show(mat_5(5))
show(mat_5(7))

matrix of size 3
    0    0    0
    0    1    0
    0    0    0

matrix of size 5
    0    0    0    0    0
    0    1    1    1    0
    0    1    2    1    0
    0    1    1    1    0
    0    0    0    0    0

matrix of size 7
    0    0    0    0    0    0    0
    0    1    1    1    1    1    0
    0    1    2    2    2    1    0
    0    1    2    3    2    1    0
    0    1    2    2    2    1    0
    0    1    1    1    1    1    0
    0    0    0    0    0    0    0



---
***Любовь и шарики***

У каждого мальчика и у каждой девочки может быть по шарику. У кого одинакового цвета -- те, наверное, влюблены. Найти все пары.

In [11]:
# Что мы знаем:
chibo = {
    "Маша": "Белый", 
    "Миша": "Зелёный",
    "Вася": "",
    "Катя": "Зелёный",
    "Ванечка": "Синий",
    "Коля": "Голубой",
    "Малика": "Фиолетовый",
    "Зоя": "Синий"
}
from pprint import pprint as pp
pp(chibo)

{'Ванечка': 'Синий',
 'Вася': '',
 'Зоя': 'Синий',
 'Катя': 'Зелёный',
 'Коля': 'Голубой',
 'Малика': 'Фиолетовый',
 'Маша': 'Белый',
 'Миша': 'Зелёный'}


In [12]:
# Найдём же все пары:
pairs = set()
for k1, v1 in chibo.items():
    for k2, v2 in chibo.items():
        if not v1 or not v2:                  # кто без шариков -- не влюблены
            continue
        if k1 != k2 and v1 == v2:             # а кто с одинаковыми - те да
            para = tuple(sorted([k1, k2]))
            pairs |= {para}
pp(pairs)

{('Катя', 'Миша'), ('Ванечка', 'Зоя')}


---
***Матрицы***

Здесь это -- двумерные массивы. В питоне размерность может быть и больше 2.

In [12]:
m = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
print(m)
from pprint import pprint as pp
pp(m)

q = [[11, 12, 13, 14, 15], 
     [21, 22, 23, 24, 25], 
     [31, 32, 33, 34, 35], 
     [41, 42, 43, 44, 45], 
     [51, 52, 53, 54, 55]]
print(q)
pp(q)

[[1, 2, 3], [4, 5, 6], [7, 8, 9]]
[[1, 2, 3], [4, 5, 6], [7, 8, 9]]
[[11, 12, 13, 14, 15], [21, 22, 23, 24, 25], [31, 32, 33, 34, 35], [41, 42, 43, 44, 45], [51, 52, 53, 54, 55]]
[[11, 12, 13, 14, 15],
 [21, 22, 23, 24, 25],
 [31, 32, 33, 34, 35],
 [41, 42, 43, 44, 45],
 [51, 52, 53, 54, 55]]


---
***Транспонирование матрицы***

Это красивое название для симметричного отражения её относительно главной диагонали.

In [13]:
from copy import deepcopy as dc
tm = dc(m)
tq = dc(q)

In [14]:
# непосредственно
for i in range(len(tm)):
    for j in range(i):
        tm[i][j], tm[j][i] = tm[j][i], tm[i][j]

pp(tm)

for i in range(len(tq)):
    for j in range(i):
        tq[i][j], tq[j][i] = tq[j][i], tq[i][j]

pp(tq)

[[1, 4, 7], [2, 5, 8], [3, 6, 9]]
[[11, 21, 31, 41, 51],
 [12, 22, 32, 42, 52],
 [13, 23, 33, 43, 53],
 [14, 24, 34, 44, 54],
 [15, 25, 35, 45, 55]]


In [15]:
# по-взрослому, спецпакетом
import numpy as np

nm = np.asmatrix(m)
pp(nm)

tnm = nm.transpose()
pp(tnm)

nq = np.asmatrix(q)
pp(nq)

tnq = nq.transpose()
pp(tnq)

matrix([[1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]])
matrix([[1, 4, 7],
        [2, 5, 8],
        [3, 6, 9]])
matrix([[11, 12, 13, 14, 15],
        [21, 22, 23, 24, 25],
        [31, 32, 33, 34, 35],
        [41, 42, 43, 44, 45],
        [51, 52, 53, 54, 55]])
matrix([[11, 21, 31, 41, 51],
        [12, 22, 32, 42, 52],
        [13, 23, 33, 43, 53],
        [14, 24, 34, 44, 54],
        [15, 25, 35, 45, 55]])


---
***Генератор речей***

In [50]:
# mk-genspeech.py 2014-05-25 1.1
# Генератор речей
# Mikhail (myke) Kolodin, mykespb@gmail.com 

import random

text = """
  Товарищи! / Господа! / Коллеги! / Дорогие товарищи! / Уважаемые коллеги! / Дамы и господа! /  Уважаемые участники! / Друзья! / Граждане соучастники! / Граждане и гражданки! / / / / / / /
  С другой стороны / Равным образом / Не следует, однако, забывать о том, что /  Таким образом / Повседневная практика показывает. что / Значимость этих проблем настолько очевидна, что / Из выступление товарища начальника ясно следует, что / В решениях съезда нашей партии указано, что/ Разнообразный и богатый опыт / Задача организации, в особенности же / Идейные соображения  высшего порядка, а также / Принимая во внимание, что / Вы, конечно, знаете, что / Вам, должно быть, известен тот факт, что
  реализация намеченных плановых заданий / рамки и место обучения кадров / постоянный количественный рост и сфера нашей активности / сложившаяся структура организации / новая модель организационной деятельности / дальнейшее развитие различных форм деятельности / постоянное информационно-пропагандистское обеспечение нашей деятельности / управление и развитие структуры / консультация с широким активом / начало повседневной работы по формированию позиции
  играет важную роль в формировании / требуют от нас анализа / требуют определения и уточнения /  способствуют подготовке и реализации / обеспечивают широкому кругу специалистов участие в формировании
  , что позволяет выполнить важные задания по разработке / , что в значительной степени обуславливает создание / , что позволяет оценить значение / , что представляет собой интересный эксперимент / , что влечёт за собой интересный процесс внедрения и модернизации
  существенных финансовых и административных условий / дальнейших направлений развития /  системы массового участия / позиций, занимаемых участниками в отношении поставленных задач /  новых предложений / направлений прогрессивного развития / форм воздействия /  модели развития / соответствующих условий активизации / системы обучения кадров, соответствующей  насущным потребностям
  """

def make_speech (rows=1):
    parts = [part.split('/') for part in text.split('\n')]
    for num in range(1, rows+1):
        print ("Пункт %d.\n" % num + "".join([random.choice(part) for part in parts]) + '.')

make_speech(4)

Пункт 1.
 Дамы и господа!  Задача организации, в особенности же  новая модель организационной деятельности  требуют определения и уточнения  , что представляет собой интересный эксперимент  соответствующих условий активизации   .
Пункт 2.
  Товарищи!  Принимая во внимание, что  начало повседневной работы по формированию позиции требуют определения и уточнения  , что позволяет оценить значение  системы обучения кадров, соответствующей  насущным потребностям  .
Пункт 3.
   Таким образом  сложившаяся структура организации  требуют от нас анализа  , что в значительной степени обуславливает создание  позиций, занимаемых участниками в отношении поставленных задач   .
Пункт 4.
 Задача организации, в особенности же  постоянный количественный рост и сфера нашей активности  требуют от нас анализа  , что влечёт за собой интересный процесс внедрения и модернизации форм воздействия   .


---
**Исключения** <a name=except></a>

---
***Деление на нуль***

In [49]:
q = 5 / 0
print(q)

ZeroDivisionError: division by zero

In [48]:
try:
    q = 5 / 0
except:
    print("тут поделили на нуль!")
    q = 0
print(q)

тут поделили на нуль!
0


In [52]:
a = 5
b = 0
# ... многобукв...

try:
    q = a / b
except:
    print("тут поделили на нуль!")
    q = 0
print(q)

тут поделили на нуль!
0


In [49]:
# практический пример полутривиальный
# пусть есть данные -- несколько списков
# надо найти для каждого некую величину, равную частному от произведения всех чисел на из сумму

data = [
    (1, 2, 3, 4, 5),
    (1, 3, 0, -2),
    (1, -1, 5, -4, -1, 0)
]

In [50]:
# неправильно
print("начинаем расчеты")
for gruppa in data:
    print(gruppa, end=" -> ")
    prod = 1
    for el in gruppa:
        prod *= el
    summa = 0
    for el in gruppa:
        summa += el
    quot = prod / summa
    print(quot)
print("продолжаем расчеты")

начинаем расчеты
(1, 2, 3, 4, 5) -> 8.0
(1, 3, 0, -2) -> 0.0
(1, -1, 5, -4, -1, 0) -> 

ZeroDivisionError: division by zero

In [51]:
# правильно
print("начинаем расчеты")
for gruppa in data:
    try:
        print(gruppa, end=" --> ")
        prod = 1
        for el in gruppa:
            prod *= el
        summa = 0
        for el in gruppa:
            summa += el
        quot = prod / summa
        print(quot)
    except ZeroDivisionError as error:
        print(error)
        quot = 0
    except e:
        print(e)
        print("something happened!")
print("продолжаем расчеты")

начинаем расчеты
(1, 2, 3, 4, 5) --> 8.0
(1, 3, 0, -2) --> 0.0
(1, -1, 5, -4, -1, 0) --> division by zero
продолжаем расчеты


---
**Математика** <a name=math></a>

---
***Линейное уравнение***

Дано: $ax + b = 0$, 
причём $a \ne 0$ и $a, b, x \in \mathbb R$. 
Требуется найти $x$ для конкретных $a, b$.

Решение: $x = \frac{-b}a$.

In [16]:
# дано: ax + b = 0
a = -3
b = 6
print(f"линейное уравнение {a}x{b:+n} = 0")

линейное уравнение -3x+6 = 0


In [17]:
# решение: x = -b / a
# если a!= 0
if a:
    x = -b / a
    print(f"решение: x = {x}")
else:
    print("а=0, это не линейное уравнение!")

решение: x = 2.0


---
***Квадратное уравнение***

Дано: $ax^2 + bx + c = 0$,
причём $a \ne 0$, и $a, b, c, x \in \mathbb R$.
Требуется найти $x$ для конкретных $a, b, c$.

Решение: $x_{1, 2} = \frac{-b \pm \sqrt{b^2 - 4ac}}{2a}$, 
если дискриминант $D=b^2-4ac > 0$, 
либо $x = \frac{-b}{2a}$, 
если $D=0$,
иначе (при $d<0$) в $\mathbb R$ решений нет.

Обозначение "$\pm$" это "плюс-минус", т.е.
$a \pm b$ обозначает сразу 2 варианта формулы: 
$a + b$ и $a - b$.

То же с индексами переменной $x$:
$x_{1, 2}$ соответствует $x_1$ и $x_2$.

Таким образом, формула выше означает:
$$
 x_1 = \frac{-b + \sqrt{b^2 - 4ac}}{2a},\,\,\,
 x_2 = \frac{-b - \sqrt{b^2 - 4ac}}{2a}.
$$

In [1]:
import math

In [2]:
# дано: a, b, c
a = 1
b = 6
c = -8
print(f"Решаем квадратное уравнение с коэффициентами {a=}, {b=}, {c=}")

Решаем квадратное уравнение с коэффициентами a=1, b=6, c=-8


In [4]:
if a:
    
    d = b**2 - 4*a*c      # это дискриминант
    print(f"дискриминант равен {d=},", 
     "положительный (2 решения)" if d>0 else
     "отрицательный (нет решений)" if d<0 else
     "нулевой (1 решение)" )

    if d > 0:
        x1 = (-b + math.sqrt(d)) / (2*a)
        x2 = (-b - math.sqrt(d)) / (2*a)
        print(f"вот 2 решения:\n{x1=}\n{x2=}")
    elif d == 0:
        x = -b / (2*a)
        print(f"вот 1 решение:\n{x=}")
    else:
        print("решений в R нет")

else:
    print("a=0, это не квадратное уравнение")

дискриминант равен d=68, положительный (2 решения)
вот 2 решения:
x1=1.1231056256176606
x2=-7.123105625617661


---
***Счастливые билеты***

Рассчитать количество счастливых билетов для номера произвольной чётной длины.

_(**Счастливым** называется билет, у которого сумма цифр левой половины равна сумме цифр правой половины,
напр., билет 123321 -- счастливый, а 123456 -- нет.)_

In [4]:
def sumdig(n):
    """сумма цифр числа
    """

    s = 0
    while n:
        s += n % 10
        n //= 10

    return s


def happy(digs = 6):
    """ считаем количество счастливых билетов
    для номера длиной digs цифр
    """

    if digs % 2:
        print("нет! Билеты с нечётной длиной номера не считаются:", digs)
        return 0

    half = digs // 2
    kol = [0 for i in range(9*half+1)]

    for n in range(10**half):
        kol[sumdig(n)] += 1

    mul = sum([n*n for n in kol])
    return mul


def org(n):
    """органайзер
    """
    print("Счастливые билеты для номеров из", n, "цифр. Их", happy(n))


def calcnums():
    """считаем всё"""

    for n in range(2, 11, 2):
        org(n)

calcnums()

Счастливые билеты для номеров из 2 цифр. Их 10
Счастливые билеты для номеров из 4 цифр. Их 670
Счастливые билеты для номеров из 6 цифр. Их 55252
Счастливые билеты для номеров из 8 цифр. Их 4816030
Счастливые билеты для номеров из 10 цифр. Их 432457640
